perf-tool/src/agent.cpp

```diff
@@ -0,0 +1,156 @@
+     * "commands" is followed by a path to the commands file
+     * "log" is followed by the path to the location for the log file
+     * "portno" is followed by a number indicating the port to run the server on 
+     */
+    while ((pos1 = oIn.find(optionsDelim)) != std::string::npos)
+    {
+        if (pos1 != std::string::npos)
+        {
+            token = oIn.substr(0, pos1);
+            if ((pos2 = token.find(pathDelim)) != std::string::npos)
+            {
+                if (!token.substr(0, pos2).compare("commandFile"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    commandsPath = token;
+                }
+                else if (!token.substr(0, pos2).compare("logFile"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    logPath = token;
+                }
+                else if (!token.substr(0, pos2).compare("portNo"))
+                {
+                    token.erase(0, pos2 + pathDelim.length());
+                    portNo = stoi(token);
+                }
+            }
+            oIn.erase(0, pos1 + optionsDelim.length());
+        }
+    }
+    if (!oIn.empty())
+    {
+        token = oIn;
+        if ((pos2 = token.find(pathDelim)) != std::string::npos)
+        {
+            if (!token.substr(0, pos2).compare("commandFile"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                commandsPath = token;
+            }
+            else if (!token.substr(0, pos2).compare("logFile"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                logPath = token;
+            }
+            else if (!token.substr(0, pos2).compare("portNo"))
+            {
+                token.erase(0, pos2 + pathDelim.length());
+                portNo = stoi(token);
+            }
+        }
+    }
+
+    printf("%s\n", commandsPath.c_str());
+    printf("%s\n", logPath.c_str());
+    printf("%i\n", portNo);
+
+    jint rest = jvm->GetEnv((void **) &jvmti, JVMTI_VERSION_1_2);
+    if (rest != JNI_OK || jvmti == NULL) {
+
+        printf("Unable to get access to JVMTI version 1.2");
+        return JNI_ERR;
+    }
+
+    jvmtiCapabilities capa;
+    jvmtiError error;
+    memset(&capa, 0, sizeof(jvmtiCapabilities));
+    capa.can_signal_thread = 1;
+    capa.can_get_owned_monitor_info = 1;
+    capa.can_generate_method_entry_events = 1;
+    capa.can_tag_objects = 1;
+    capa.can_get_current_thread_cpu_time = 1;
+    capa.can_get_line_numbers = 1;
+    capa.can_generate_vm_object_alloc_events = 1;
+    capa.can_generate_monitor_events = 1;
+    capa.can_generate_exception_events = 1;
+    capa.can_get_source_file_name = 1;
+    error = jvmti->AddCapabilities(&capa);
+    check_jvmti_error(jvmti, error, "Failed to set jvmtiCapabilities.");
+
+    error = jvmti->SetEventNotificationMode(JVMTI_ENABLE, JVMTI_EVENT_VM_INIT, (jthread)NULL);
```

Comment:
It's OK to terminate the JVM in this code. If the JVMTI_EVENT_VM_INIT cannot be hooked, then the `Server` object is never created and we'll crash later on when trying to use it.